### PR TITLE
Fix pg_basebackup with tablespace when using --target-gp-dbid.

### PR DESCRIPTION
**Pull request by berlin-ab**
During pg_basebackup's initial setup it verifies that the tablespace that it will write into exists and is empty. Later, while receiving and unpacking tar files, it tries to create a directory for the file that it is unpacking. In this situation, pg_basebackup errors out because the directory already exists.

Our change checks to see if the directory that it is trying to create during the unpacking is one that we've already created. If so, it continues to the next file, avoiding the error.

### CODE REVIEW

File: src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+-- And a database using the tablespace
+drop database if exists some_database_with_tablespace;
+create database some_database_with_tablespace tablespace some_isolation2_pg_basebackup_tablespace;
+
+-- When we create a full backup
+select pg_basebackup(address, 100, port, 'some_replication_slot', '/tmp/some_isolation2_pg_basebackup') from gp_segment_configuration where content = 0 and role = 'p';
+
+-- Then we should have a backup of the source segment files in the newly created target tablespace
+-- expect: 1
```
Comment (ashwinstar) [inline src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql]: If we are keeping this code then should explain Why expect to see 1, which is for the only database directory which should get created here.

File: src/bin/pg_basebackup/pg_basebackup.c [hunk 2]
```diff
@@ -1303,9 +1311,14 @@ ReceiveAndUnpackTarFile(PGconn *conn, PGresult *res, int rownum)
 							continue;
 
 						rmtree(filename, true);
+
 					}
 
-					if (mkdir(filename, S_IRWXU) != 0)
+					bool is_gp_tablespace_directory = strcmp(gp_tablespace_filename, filename) == 0;
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/bin/pg_basebackup/pg_basebackup.c]: Variables like these help me understand what's going on quicker.

File: src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql [hunk 3]
```diff
@@ -0,0 +1,22 @@
+include: helpers/gp_management_utils_helpers.sql;
+
+-- Ensure we have a clean slate to begin with
+!\retcode rm -rf /tmp/some_isolation2_pg_basebackup;
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql]: I find the template files extremely hard to read, especially when syntax highlighting from the editor gets confused.

**Comment by ashwinstar** [comment]
On Mon, Feb 4, 2019 at 11:39 AM Adam Berlin <notifications@github.com>
wrote:

> *@berlin-ab* commented on this pull request.
> ------------------------------
>
> In src/bin/pg_basebackup/pg_basebackup.c
> <https://github.com/greenplum-db/gpdb/pull/6889#discussion_r253615102>:
>
> >  					}
>
> -					if (mkdir(filename, S_IRWXU) != 0)
> +					bool is_gp_tablespace_directory = strcmp(gp_tablespace_filename, filename) == 0;
>
> Variables like these help me understand what's going on quicker.
>
Seems unnecessary work generated for compiler to me. If the purpose is just
to ease understanding, code comment serves the purpose better than just
having variable for the same as can even express why and not just what.

**Comment by ashwinstar** [comment]
On Mon, Feb 4, 2019 at 12:31 PM Adam Berlin <notifications@github.com>
wrote:

> *@berlin-ab* commented on this pull request.
> ------------------------------
>
> In src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql
> <https://github.com/greenplum-db/gpdb/pull/6889#discussion_r253634265>:
>
> > @@ -0,0 +1,22 @@
> +include: helpers/gp_management_utils_helpers.sql;
> +
> +-- Ensure we have a clean slate to begin with
> +!\retcode rm -rf /tmp/some_isolation2_pg_basebackup;
>
> I find the template files extremely hard to read, especially when syntax
> highlighting from the editor gets confused.
>
If that's issue we should fix the same, by teaching the editor to interpret
.source similar to .sql. As irrespective template piece is not going away
and used in upstream and is super helpful for such kind of things.

File: src/bin/pg_basebackup/pg_basebackup.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/bin/pg_basebackup/pg_basebackup.c]: Thanks, nice catch. I'll get that in my muscle memory sometime soon.

File: src/bin/pg_basebackup/pg_basebackup.c [hunk 4]
```diff
@@ -1303,9 +1311,14 @@ ReceiveAndUnpackTarFile(PGconn *conn, PGresult *res, int rownum)
...
+					if (is_gp_tablespace_directory) {
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/bin/pg_basebackup/pg_basebackup.c]: I kept the braces, as I added a comment to this section.

File: src/bin/pg_basebackup/pg_basebackup.c [hunk 5]
```diff
@@ -1303,9 +1311,14 @@ ReceiveAndUnpackTarFile(PGconn *conn, PGresult *res, int rownum)
...
+						continue;
+					}
+					else if (mkdir(filename, S_IRWXU) != 0)
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/bin/pg_basebackup/pg_basebackup.c]: Nice catch.

File: src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql]: @ashwinstar I'd like to improve this by introducing the templates, but first I'd like to understand how using a template helps. Do we have people using a different location for `/tmp`?

[the remark this answers is not in the document]
Comment (jchampio) [reply src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql]: Drive-by comment -- I don't think this is necessarily a blocker for *this* PR -- but how much of this test would have been implemented in the pg_basebackup TAP tests if we had those running correctly?

The question I'm getting at is, is it worth getting perfect now, or should we focus on coverage in the short term and migration to TAP in the medium term?

[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/test/isolation2/sql/segwalrep/pg_basebackup_with_tablespaces.sql]: Good point. I hadn't really thought about it. I think the framing of coverage now, TAP later is a good plan though.

**Comment by berlin-ab** [comment]
> If that's issue we should fix the same, by teaching the editor to interpret .source similar to .sql. As irrespective template piece is not going away and used in upstream and is super helpful for such kind of things.

Good point. Here's an example of how to do that in emacs:

https://www.emacswiki.org/emacs/AutoModeAlist

And here's the example in the wild from my emacs config:

https://github.com/berlin-ab/gpdb-tools/blob/6a6241f0e30933c1f0685eda6201865b75b414f9/emacs-config.el#L108

**Comment by berlin-ab** [comment]
Legit failure in `gpcheckcat`. Fixing now.

**Comment by berlin-ab** [comment]
Found the cause of the `gpcheckcat` failure. Going to push after the PR pipeline goes green.

**Pull request merged**